### PR TITLE
Link filter refactoring

### PR DESCRIPTION
**Pull request by ArthurHoaro**
* introduce class LinkFilter to handle link filter operation (and lighten LinkDB).
  * handle 'private only' in filtering.
  * update template to prefill search fields with current search terms.
  * coding style.
  * unit test (mostly move from LinkDB to LinkFilter).

PS: preparation for #358 #315 and 'AND' search.

### CODE REVIEW

**Comment by ArthurHoaro** [comment]
I have also included a correction for #426 (blank tags).

File: index.php [hunk 1]
```diff
@@ -730,18 +731,23 @@ function showRSS()
...
+        $linksToDisplay = $LINKSDB;
+    }
 
     $nblinksToDisplay = 50;  // Number of links to display.
-    if (!empty($_GET['nb']))  // In URL, you can specificy the number of links. Example: nb=200 or nb=all for all links.
-    {
-        $nblinksToDisplay = $_GET['nb']=='all' ? count($linksToDisplay) : max($_GET['nb']+0,1) ;
+    // In URL, you can specificy the number of links. Example: nb=200 or nb=all for all links.
+    if (!empty($_GET['nb'])) {
+        $nblinksToDisplay = $_GET['nb'] == 'all' ? count($linksToDisplay) : max($_GET['nb']+0, 1);
```
Comment (virtualtam) [inline index.php]: is `$_GET['nb']+0` used to ensure the value is cast to a (positive) integer?
if so, could we add an explicit string to int conversion, e.g. with [`intval()`](http://php.net/manual/en/function.intval.php)?

File: index.php [hunk 2]
```diff
@@ -1283,12 +1298,14 @@ function renderPage()
 
         if (isset($params['searchtags'])) {
             $tags = explode(' ',$params['searchtags']);
-            $tags=array_diff($tags, array($_GET['removetag'])); // Remove value from array $tags.
-            if (count($tags)==0) {
+            // Remove value from array $tags.
+            $tags = array_diff($tags, array($_GET['removetag']));
+            $params['searchtags'] = implode(' ',$tags);
```
Comment (virtualtam) [inline index.php]: +space after comma

File: index.php [hunk 3]
```diff
@@ -1865,81 +1884,78 @@ function importFile()
...
+    // Search by tag
+    elseif (isset($_GET['searchtags'])) {
+        $search_crits = explode(' ', escape(trim($_GET['searchtags'])));
+        $search_type = LinkFilter::$FILTER_TAG;
+        $linksToDisplay = $LINKSDB->filter($search_type, $search_crits, false, $privateonly);
+    }
+    // Detect smallHashes in URL.
+    elseif (isset($_SERVER['QUERY_STRING'])
+        && preg_match('/[a-zA-Z0-9-_@]{6}(&.+?)?/', $_SERVER['QUERY_STRING'])) {
+        $search_type = LinkFilter::$FILTER_HASH;
```
Comment (virtualtam) [inline index.php]: `LinkFilter::$FILTER_HASH` can be passed directly to the call to `$LINKSDB->filter(...)`

File: index.php [hunk 4]
```diff
@@ -1865,81 +1884,78 @@ function importFile()
...
     }
 
     // ---- Handle paging.
-    /* Can someone explain to me why you get the following error when using array_keys() on an object which implements the interface ArrayAccess???
-       "Warning: array_keys() expects parameter 1 to be array, object given in ... "
-       If my class implements ArrayAccess, why won't array_keys() accept it ?  ( $keys=array_keys($linksToDisplay); )
-    */
-    $keys=array(); foreach($linksToDisplay as $key=>$value) { $keys[]=$key; } // Stupid and ugly. Thanks PHP.
+    $keys = array();
+    foreach($linksToDisplay as $key => $value) {
```
Comment (virtualtam) [inline index.php]: +space after `foreach`

File: index.php [hunk 5]
```diff
@@ -1865,81 +1884,78 @@ function importFile()
...
+    // Detect smallHashes in URL.
+    elseif (isset($_SERVER['QUERY_STRING'])
+        && preg_match('/[a-zA-Z0-9-_@]{6}(&.+?)?/', $_SERVER['QUERY_STRING'])) {
+        $search_type = LinkFilter::$FILTER_HASH;
+        $search_crits = substr(trim($_SERVER["QUERY_STRING"], '/'), 0, 6);
+        $linksToDisplay = $LINKSDB->filter($search_type, $search_crits);
+
+        if (count($linksToDisplay) == 0) {
+            header($_SERVER['SERVER_PROTOCOL'] . ' 404 Not Found');
+            echo '<h1>404 Not found.</h1>Oh crap.
```
Comment (virtualtam) [inline index.php]: #430 raised to introduce a proper `404` template :)

**Comment by virtualtam** [comment]
Nice work!
Maybe for contributions that big, the additional code-cleanup bits should be part of a dedicated commit, this would help a lot when reviewing the functional changes :)

**Comment by ArthurHoaro** [comment]
Thanks! Yep, I'll try to do it next time. It's not really easy since I tend to format the code when I touch it.

File: index.php (hunk 1 above)
Comment (ArthurHoaro) [inline index.php]: Done. However, `intval()` returns 0 on failure. ಠ_ಠ
That's why the `max()` function is here.

File: index.php (hunk 3 above)
Comment (ArthurHoaro) [inline index.php]: Yes, but no. It's assign later as a template variable.

File: index.php (hunk 5 above)
Comment (ArthurHoaro) [inline index.php]: Good idea!

**Comment by ArthurHoaro** [comment]
To template maintainers, improvements in `linklist.html` [L14](https://github.com/ArthurHoaro/Shaarli/commit/822bffced8212e7f34bcb2ad063b31a78bd57bdb#diff-ce2b0a75ea8d4627a55eb34af413caa9R14), [L22](https://github.com/ArthurHoaro/Shaarli/commit/822bffced8212e7f34bcb2ad063b31a78bd57bdb#diff-ce2b0a75ea8d4627a55eb34af413caa9R14) and [L56](https://github.com/shaarli/Shaarli/commit/eefb636cea7acef9ddfd02a90749820f5fafc9f6#diff-ce2b0a75ea8d4627a55eb34af413caa9R56).

**Pull request merged**